lighthouse-core/audits/polyfills.js

```diff
@@ -0,0 +1,225 @@
+      let subpattern = '';
+
+      // String.prototype.startsWith =
+      subpattern += `${object || ''}\\.?${property}\\s*=`;
+      subpattern += `|${object || ''}\\[${qt(property)}\\]\\s*=`;
+
+      // Object.defineProperty(String.prototype, 'startsWith'
+      subpattern += `|defineProperty\\(${object || qt('window')},\\s*${qt(property)}`;
+
+      // check for lodash. Assume that there is a lodash module for most polys.
+      // TODO this would only work if the bundle has not been stripped of module names (and replaced with an index into the bundled modules),
+      // which is unlikely in a production environment. Instead, look at source map?
+      // TODO how to detect if all of lodash is bundled (a waste)?
+      // TODO maybe just punt entirely for now?
+      // subpattern += `|lodash/${property}`;
+
+      return `(${subpattern})`;
+    }).join('|');
+
+    const polysSeen = new Set();
+    const polyMatches = [];
+    const re = new RegExp(`(^\r\n|\r|\n)|${pattern}`, 'g');
+    /** @type {RegExpExecArray | null} */
+    let result;
+    let row = 0;
+    let rowBeginsAtIndex = 0;
+    // each poly maps to one subgroup in the generated regex. for each iteration of the regex exec,
+    // only one subgroup will be defined. Exec until no more matches.
+    while ((result = re.exec(code)) !== null) {
+      // discard first (it's the whole matching pattern)
+      // index 1 is truthy if matching a newline, and is used to track the row number
+      // matches maps to each possible poly. Exec until no more matches
+      // only one of [isNewline, ...matches] is ever defined.
+      const [, isNewline, ...matches] = result;
+      if (isNewline) {
+        row++;
+        rowBeginsAtIndex = result.index;
+        continue;
+      }
+      const poly = polys[matches.findIndex(Boolean)];
+
+      // don't report more than one instance of a poly for this code.
+      // would be nice, but it also reports false positives if both '='
+      // and 'Object.defineProperty' are used conditionally based on feature detection.
+      if (polysSeen.has(poly)) {
+        continue;
+      }
+
+      polysSeen.add(poly);
+      polyMatches.push({
+        row,
+        col: result.index - rowBeginsAtIndex,
+        poly,
+      });
+    }
+
+    return polyMatches;
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts, context) {
+    const devtoolsLog = artifacts.devtoolsLogs[Polyfills.DEFAULT_PASS];
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+
+    // TODO
+    // how do we determine which polys are not necessary?
+    // ex: only reason to polyfill Array.prototype.filter is if IE <9,
+    //     only reason to polyfill String.prototype.startsWith is if IE, ...
+    // Is there a standard way to declare "we support these browsers"? A meta tag?
+    /** @type {Poly[]} */
+    const polys = [
```

Comment:
maybe we could pull some of this code out of the main `audit` function and break it down a bit?